hawkbit-artifact-repository-filesystem/src/main/java/org/eclipse/hawkbit/artifact/repository/ArtifactFilesystemRepository.java

```diff
@@ -0,0 +1,188 @@
+        return renameFileToSHA1Naming(file, artifact);
+    }
+
+    @Override
+    public void deleteBySha1(final String sha1Hash) {
+        FileUtils.deleteQuietly(getFile(sha1Hash));
+    }
+
+    @Override
+    public ArtifactFilesystem getArtifactBySha1(final String sha1) {
+        final File file = getFile(sha1);
+        if (!file.exists()) {
+            return null;
+        }
+        final ArtifactFilesystem artifact = new ArtifactFilesystem(file);
+        artifact.setArtifactId(sha1);
+        artifact.setHashes(new DbArtifactHash(sha1, null));
+        artifact.setSize(file.length());
+        return artifact;
+    }
+
+    private DbArtifact store(final InputStream content, final String contentType, final DbArtifactHash hash,
+            final MessageDigest mdSHA1, final MessageDigest mdMD5, final File file) {
+        final DbArtifact artifact = new DbArtifact();
+        try (final DigestOutputStream outputstream = openFileOutputStream(file, mdSHA1, mdMD5)) {
+            final long artifactSize = ByteStreams.copy(content, outputstream);
+            outputstream.flush();
+            final String sha1Hash = BaseEncoding.base16().lowerCase().encode(mdSHA1.digest());
+            final String md5Hash = BaseEncoding.base16().lowerCase().encode(mdMD5.digest());
+            artifact.setArtifactId(sha1Hash);
+            artifact.setSize(artifactSize);
+            artifact.setContentType(contentType);
+            artifact.setHashes(new DbArtifactHash(sha1Hash, md5Hash));
+            checkHashes(artifact, hash);
+        } catch (final IOException e) {
+            throw new ArtifactStoreException(e.getMessage(), e);
+        } catch (final HashNotMatchException e) {
+            file.delete();
+            throw e;
+        }
+        return artifact;
+    }
+
+    private ArtifactFilesystem renameFileToSHA1Naming(final File file, final DbArtifact artifact) {
+        final File fileSHA1Naming = getFile(artifact.getHashes().getSha1());
+        final ArtifactFilesystem fileSystemArtifact = new ArtifactFilesystem(fileSHA1Naming);
+        if (fileSHA1Naming.exists()) {
+            FileUtils.deleteQuietly(file);
+        } else if (!file.renameTo(fileSHA1Naming)) {
+            throw new ArtifactStoreException("Could not store the file " + fileSHA1Naming);
+        }
+
+        file.delete();
+        fileSystemArtifact.setArtifactId(artifact.getArtifactId());
+        fileSystemArtifact.setContentType(artifact.getContentType());
+        fileSystemArtifact.setHashes(artifact.getHashes());
+        fileSystemArtifact.setSize(artifact.getSize());
+        return fileSystemArtifact;
+    }
+
+    private DbArtifact checkHashes(final DbArtifact artifact, final DbArtifactHash hash) {
+        if (hash == null) {
+            return artifact;
+        }
+        if (hash.getSha1() != null && !artifact.getHashes().getSha1().equals(hash.getSha1())) {
```

Comment:
yeah `hash.getSha1()` and `hash.getMd5()` can be null. There will be `null` if no SHA1 hash is provided. 

`JpaArtifactManagement`
```
try {
            result = artifactRepository.store(stream, filename, contentType,
                    new DbArtifactHash(providedSha1Sum, providedMd5Sum));
        }
```